### ISSUE TITLE
Field already defined

### ISSUE THREAD

**Issue by tyers** [jpmml/sklearn2pmml#189]
Hi Villu,
Not an issue as such, but i'm struggling with some of my defined features and i'm at a loss as to how to avoid a field already defined error when converting to pmml.

One of the features i am generating is the age of an individual, represented as follows in a dataframemapper:
```
(['CURRENT_DATE', 'DATE_OF_BIRTH'],
[
    DateDomain(invalid_value_replacement=None, invalid_value_treatment='as_missing', missing_value_replacement='2010-01-01 00:00:00', missing_value_treatment='as_value'),
    DaysSinceYearTransformer(year=2010),
    ExpressionTransformer(expr='(X[0] - X[1]) / 365'),
    SimpleImputer(fill_value=-1, strategy='constant')
],
{'alias': 'AGE'})
```
This converts fine to pmml in isolation, however, if i would like to calculate any other date difference using either of those dates i get the error reported upon conversion. eg:
```
(['CURRENT_DATE', 'DATE_OF_BIRTH_2'],
[ ... identical transformation steps ... ],
{'alias': 'AGE_2'})
```
would result in the following error:
`java.lang.IllegalArgumentException: Field days_since_year(CURRENT_DATE) is already defined.`

As i'm passing a 2D array in to the transformations i cannot see a way to wrap these up with an Alias decorator to ensure that there is not conflict.

In this instance the situation is that the derived feature that is being generated in the pmml is actually an exact duplicate, so not sure if the conversion could detect that and re-use the existing definition?

Thanks

**Comment by vruusmann** [comment]
> the derived feature that is being generated in the pmml is actually an exact duplicate, so not sure if the conversion could detect that

The conversion library checks for duplicate field definitions with some other transformer classes, but not with `DurationTransformer` subclasses.

However, in order to be able to detect exact duplicates, then the generated field name must include the reference date as well:
```
days_since_year(CURRENT_DATE, 1970)
days_since_year(CURRENT_DATE, 1980)
```

**Comment by vruusmann** [comment]
It should be possible to solve duplicate field exceptions by rearranging the pipeline.

The idea is to have two `sklearn.compose.ColumnTransformer` (or plain old `DataFrameMapper`) instances chained one after another. The first one derives `days_since_year(CURRENT_DATE)`. The second one can then reference it by name, and generate two or more derived fields based on it.

This trick would be easier to implement using `ColumnTransformer`, because it provides a nice way of passing on untransformer columns. For example, you could derive `days_since_year(CURRENT_DATE)`, and simply "passthrough" all other columns to the next transformer battery.

**Comment by tyers** [comment]
hmm ok. I was worried that something like that was going to be the only solution.
i've so far avoided using a column transformer as i'm not aware of a robust method to access the names of the transformed features for use in assessing the feature importance, and doing model explanation etc because it requires the use of sklear.Pipeline objects to chain multiple transformers, and these do not implement a get_feature_names method.

**Comment by tyers** [comment]
hi villu,
update on the above - turns out that DataFrameMapper also has a passthrough option implemented via the default=None property value.
However this is not supported by jpmml-sklearn with the following error:
```
java.lang.IllegalArgumentException: Attribute 'sklearn_pandas.dataframe_mapper.DataFrameMapper.default' has a missing (None/null) value
        at org.jpmml.sklearn.PyClassDict.get(PyClassDict.java:46)
        at org.jpmml.sklearn.PyClassDict.getObject(PyClassDict.java:97)
        at sklearn_pandas.DataFrameMapper.getDefault(DataFrameMapper.java:97)
        at sklearn_pandas.DataFrameMapper.initializeFeatures(DataFrameMapper.java:47)
        at sklearn.Initializer.encodeFeatures(Initializer.java:41)
        at sklearn.Transformer.updateAndEncodeFeatures(Transformer.java:85)
        at sklearn.pipeline.Pipeline.encodeFeatures(Pipeline.java:83)
        at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:203)
        at org.jpmml.sklearn.Main.run(Main.java:145)
        at org.jpmml.sklearn.Main.main(Main.java:94)
```
would it be possible for this to be supported via an enhancement given the equivalent passthrough option is supported in the ColumnTransformer class?

**Comment by vruusmann** [comment]
> turns out that DataFrameMapper also has a passthrough option implemented via the default=None property value.

It's impossible to support `DataFrameMapper.default = None` (it has been requested before, see closed issues).

In brief, `ColumnTransformer` retains the name(s) of passed-through columns, which allows the JPMML-SkLearn library to correctly determine the "dimensionality" of input data. `DataFrameMapper.default = None` doesn't.
